File: backend/src/main/kotlin/com/swpp/footprinter/domain/trace/model/Trace.kt
package com.swpp.footprinter.domain.trace.model

import com.swpp.footprinter.common.model.BaseEntity
import com.swpp.footprinter.common.utils.ImageUrlUtil
import com.swpp.footprinter.domain.footprint.model.Footprint
import com.swpp.footprinter.domain.trace.dto.TraceDetailResponse
import com.swpp.footprinter.domain.trace.dto.TraceResponse
import com.swpp.footprinter.domain.user.model.User
import javax.persistence.*

@Entity
class Trace(

    @Column(name = "trace_title", nullable = false)
    val traceTitle: String,

    @Column(name = "trace_date", nullable = false)
    val traceDate: String,

    @Column(name = "public", nullable = false)
    var isPublic: Boolean = true,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(referencedColumnName = "id", name = "userId")
    val owner: User,

    @OneToMany(mappedBy = "trace", cascade = [CascadeType.ALL])
    val footprints: MutableSet<Footprint>,

    @Column(name = "likes_count", columnDefinition = "integer default 0")
    var likesCount: Int = 0,

    @Column(name = "view_count", columnDefinition = "integer default 0")
    var viewCount: Int = 0,

) : BaseEntity() {
    fun toResponse(imageUrlUtil: ImageUrlUtil): TraceResponse {
        return TraceResponse(
            id = id,
            date = traceDate,
            title = traceTitle,
            owner = owner.toResponse(imageUrlUtil),
            likesCount = likesCount
        )
    }

    fun toDetailResponse(imageUrlUtil: ImageUrlUtil, isLiked: Boolean = false): TraceDetailResponse {
        return TraceDetailResponse(
            id = id,
            date = traceDate,
            title = traceTitle,
            owner = owner.toResponse(imageUrlUtil),
            isLiked = isLiked,
            footprints = footprints.map { it.toResponse(imageUrlUtil) }.sortedBy { it.startTime },
            likesCount = likesCount,
            viewCount = viewCount,
            isPublic = isPublic
        )
    }
}

File: frontend/dto/trace.ts
import { FootprintRequestType, FootprintResponseType } from "./footprint";
import { UserResponseType } from "./user";

export interface TraceRequestType {
  title: string;
  // date: string;
  public: boolean;
  footprintList: FootprintRequestType[];
}

export interface TraceDetailResponseType {
  id: number;
  date: string;
  title: string;
  owner: UserResponseType;
  likesCount: number;
  isLiked: boolean;
  footprints: FootprintResponseType[];
  viewCount: number;
  isPublic: boolean;
}

export interface TraceViewResponseType {
  viewCount: number;
}

export interface TraceLikeResponseType {
  likesCount: number;
  isLiked: boolean;
}

export interface TraceViewResponseType {
  viewCount: number;
}
